pennylane/vqe/vqe.py

```diff
@@ -451,6 +526,12 @@ def metric_tensor(self, args, kwargs=None, diag_approx=False, only_construct=Fal
         Returns:
             array[float]: metric tensor
         """
+        if self._optimize:
+            raise ValueError(
+                "Evaluation of the metric tensor is not supported when using "
+                "optimized observables. Set the argument optimize=False to obtain "
+                "the metric tensor."
+            )
```

Comment:
Hey @trbromley, I thought about this more, and I don't see why it shouldn't work in practice. The Fubini-Study metric tensor is independent of the variational circuit measured observables --- it only depends on the circuit ansatz.

So what we could do here:

* Construct a QNode that _doesn't_ use tape mode, which simply has the ansatz followed by a trivial PauliZ observable.
* Wrap this QNode in a wrapper function to return `qnode.metric_tensor(args=args, kwargs=kwargs, diag_approx=diag_approx)`.

As long as PennyLane lets us freely mix tape mode and non-tape mode, this should work and be very simple to code up.

Reply:
Note: this is a short term fix:

* It is not trivial to support QNG in tape mode, so likely this can't be done over the next few days

* However, if using VQECost, we could behind-the-scenes simply uses non-tape mode for the Fubini-Study metric tensor, and tape mode for the cost function and gradients.

The longer term fix would be to support the metric tensor calculation in tape mode

Reply:
Good idea, I'm having a go in this branch but having some troubles: https://github.com/PennyLaneAI/pennylane/tree/ch2397-metric

I think we should merge this in now and follow up if we can get metric tensor support.